design/ip-address-management-for-networkdata.md

```diff
@@ -0,0 +1,328 @@
+  that will not be used. This will result in large gaps and inefficiency in the
+  addresses allocation, specifically if those are public IPv4 addresses.
+* It is not possible to rule out some IP addresses that might be in use for any
+  reason after the deployment.
+
+### Goals
+
+- Introduce a Metal3IPPool and a Metal3IPAddress CRDs.
+- add flexibility in the IP address allocation
+- enable sharing of a pool across machine deployments / KCP
+- enable use of non-continuous pools
+- enable external IP management by using Metal3IPAddress CRs
+- keep when possible the predictability of the IP address based on the
+  Metal3Data index. When shared among multiple Metal3DataTemplate, the
+  allocation will happen on a first come first served basis.
+
+### Non-Goals
+
+- provide a DHCP feature or any equivalent level feature.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a user deploying a target Kubernetes cluster, I want to be able to specify
+a non-continuous range of public IP addresses to be used for the nodes static
+allocation as those sparse addresses are all I got from my network admins.
+
+#### Story 2
+
+As a user deploying a target Kubernetes cluster, I want to be able to specify
+the pool of public IP addresses available once, and have any machine deployment
+or kubeadm control plane referencing it use it, effectively sharing it among
+multiple machine deployments.
+
+#### Story 3
+
+As a user deploying a target Kubernetes cluster, I want to keep the allocation
+predictability offered by the current feature, i.e. the ip address is the
+start address + the index of the Metal3Data, when I use a pool for a single
+machine deployment. This is to ensure that the current proposal adds a feature
+without removing any existing feature and is fully backward compatible.
+
+### Implementation Details/Notes/Constraints
+
+- This proposal should be fully backwards compatible and not modify any existing
+  behaviour.
+- When shared among multiple machine deployment, the allocation would be
+  random as much as possible to avoid conflicts and re-use as much as possible.
+- If not needed, this feature should not require the user to modify anything
+  existing in their CRs
+
+### Risks and Mitigations
+
+This adds some complexity to the Metal3 system. Proper documentation could solve
+the issue, and for users not needing this feature, no change would be required.
+
+## Design Details
+
+
+A Metal3IPPool object would be created :
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3IPPool
+metadata:
+  name: pool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Cluster
+    name: cluster-1
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    kind: Metal3Data
+    name: metal3data-1
+spec:
+  pools:
+    - start: 192.168.0.10
+      end: 192.168.0.15
+      subnet: 192.168.0.0/24
+      step: 1
```

Comment:
your pools should not overlap. Maybe I need to add this in the proposal. It is a requirement that no pool overlap with any other in the target cluster being deployed. There should be no conflict between the pools given by the user. We can validate it within the same Metal3IPPool, but I don't see how to do it across Metal3IPPools.

Reply:
And about the step value, I thought it could be used for IPv6, I don't yet see a use-case, so maybe I should just remove it ? It was mostly for feature parity with the previous proposal.

Reply:
Yeah, let's leave it out for now. We can add it later if we change our minds.

Reply:
ok, I'll remove it from the proposal